### DISCUSSION TITLE
GTFS import for transit

### DISCUSSION THREAD

**Discussion by pranavpandey1998official** [valhalla/valhalla#2757] [Ideas]
Hi, I am trying to add GTFS import, so let's discuss 

Ok so the main problem I am facing is that transitland stops have a osm_way_id property (https://www.transit.land/documentation/datastore/stops), which is used in protobuf node, we cannot have that from gtfs, and because the gtfs is imported before the tiles are built I don't think I can do fuzzy matching , its a chicken and an egg problem, also I tried reading transitBuilder and I think that property is very important for building the transit tiles,

Also there is a problem with onestopID (GTFS doesn't have that) we can use the algo that transitland is using to generate that or do you want to just randomly give them
@kevinkreiser

**Comment by kevinkreiser** [comment]
i propose the following:

1. move the transit builder stage of tile building to be a post processing step similar to how we add historical traffic after the fact
2. leave the osmway id blank in the created transit protobuf
3. in the new post processing stage of transit building, if the way id is missing, we'll fall back to finding an appropriate way to connect the transit station to on-the-fly. this may not always be the best place to connect it and it certainly wont be a user-curated way to connect it but it should function well enough.

So i guess what im proposing to you specifically is  dont worry about getting the way_id in there for now, i can help with 1 and 3 above once you have a process for creating transit pbfs.

**Comment by pranavpandey1998official** [comment]
@kevinkreiser hi, I want to use this for parsing the gtfs,(https://github.com/mapsme/just_gtfs) its going to make my life a lot easier as reading CSV , and coding considering all GFTS nuances (like optional feild and missing some id etc) is pain in ass,

few points its using c++17 and we are at c++11 , i find out the using boost::optional in place of std::option and removing all compiler warning related to c++17 usage like ` 'if' initialization statements` and `inline varialbles` I can make it work , so is it cool with you guys

**Reply by mandeepsandhu** [reply]
I don't think Valhalla uses c++17. Where do you see the c++17 specific syntax?

**Reply by nilsnolde** [reply]
other way around, `just_gtfs` uses c++17.. pity though. header-only and not having to care about gtfs stuff would be really nice! valhalla is at c++14. not sure what the plans/effort are to move one higher..

**Reply by kevinkreiser** [reply]
correct we wont be able to make use of that header unless we upgrade to c++17 which we cant do at the moment im afraid @pranavpandey1998official

**Reply by nilsnolde** [reply]
But how about using a c++14 adapted version of it @kevinkreiser ? Would you consider taking that in as a “fork” for being able to handle gtfs?

**Reply by kevinkreiser** [reply]
do you have some interest to fork it into the valhalla repo and port it? then we can vendor it into valhalla from the fork

**Reply by nilsnolde** [reply]
Yeah that’s what @pranavpandey1998official was thinking I guess. But cool that you’d accept! Should be minor effort and add more value than cursed CI crap😅so I could try to help out a little if you want @pranavpandey1998official

**Reply by pranavpandey1998official** [reply]
@nilsnolde that would be awesome, like I said I can make it work with minor changes, and its working fine till now. I am moving forward with my version of that library to build on top of it, tell me when you are done with it thanks

**Reply by nilsnolde** [reply]
Ah great you made it work already! Eventually that project would need to live as a sub project in third_party, like the other vendored packages do. Easiest is you push the c++14 version into its own repo.

**Reply by pranavpandey1998official** [reply]
https://github.com/pranavpandey1998official/just_gtfs this this the fork with changes I have made, If you wanna look into it

**Reply by mandeepsandhu** [reply]
why not upstream the C++14 support? maybe check with the original author to see if they'd be willing to accept patches for it. IF they aren't willing to accept the patch, then forking it is fine. Just my 2¢.

**Reply by pranavpandey1998official** [reply]
I was able to port it using boost, I don't think he/she will be able to accept that as external dependency, We can request them to accept it as a new branch , I think that could be possible

**Comment by pranavpandey1998official** [comment]
hi @kevinkreiser, I want to know how you are handling (transit) route which lies in different tiles, let say you have a route which extends 2 tiles, are you going to include it in both tiles? 
vallahala_fetch_transit is requesting the route from transit land and passing the bbox as query parameter , I am not sure what is returned.

Also If I include a route in a tile in which it doesn't belong is that ok? (ps - I will just include all the routes in all the tiles)
(here route means the proto message route)

**Reply by kevinkreiser** [reply]
The routes are stored in the tiles they begin in and not other tiles. They should not be stored in all tiles

**Comment by pranavpandey1998official** [comment]
hi @kevinkreiser , I am having trouble understanding the frequency based transit in transit land
```
(taken from https://v1.transit.land/documentation/datastore/schedules.html)
frequency_headway_seconds | Integer | Frequency-based schedule headway, in seconds
-- | -- | --
frequency_start_time | Time | Start of frequency-based schedule
frequency_end_time | Time | End of frequency-based schedule
```

this is what is given in the docs of transit land, here frequency_start_time would be the time at which the frequency star
ts for the whole trip,(like in gtfs) or the time at which frequency stats at the origin of the stop pair ?

**Reply by pranavpandey1998official** [reply]
hi @kevinkreiser  any updates?

**Comment by pranavpandey1998official** [comment]
here is the PR if anyone want to give suggestion https://github.com/valhalla/valhalla/pull/2835 ✌️
